src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -0,0 +1,115 @@
+package com.pusher.client.channel.impl;
+
+import com.pusher.client.AuthorizationFailureException;
+import com.pusher.client.Authorizer;
+import com.pusher.client.channel.ChannelState;
+import com.pusher.client.channel.PrivateEncryptedChannel;
+import com.pusher.client.channel.PrivateEncryptedChannelEventListener;
+import com.pusher.client.channel.SubscriptionEventListener;
+import com.pusher.client.connection.impl.InternalConnection;
+import com.pusher.client.crypto.nacl.SecretBoxOpener;
+import com.pusher.client.util.Factory;
+import com.pusher.client.util.internal.Base64;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class PrivateEncryptedChannelImpl extends ChannelImpl implements PrivateEncryptedChannel {
+
+    private final InternalConnection connection;
+    private final Authorizer authorizer;
+    protected SecretBoxOpener secretBoxOpener;
+
+    public PrivateEncryptedChannelImpl(final InternalConnection connection,
+                                       final String channelName,
+                                       final Authorizer authorizer,
+                                       final Factory factory) {
+        super(channelName, factory);
+        this.connection = connection;
+        this.authorizer = authorizer;
+    }
+
+    @Override
+    public void bind(final String eventName, final SubscriptionEventListener listener) {
+
+        if (!(listener instanceof PrivateEncryptedChannelEventListener)) {
+            throw new IllegalArgumentException(
+                    "Only instances of PrivateEncryptedChannelEventListener can be bound " +
+                            "to a private encrypted channel");
+        }
+
+        super.bind(eventName, listener);
+    }
+
+    private byte[] authenticate() {
+
+        try {
+            final Map authResponseMap = GSON.fromJson(getAuthResponse(), Map.class);
+            final String auth = (String) authResponseMap.get("auth");
+            final String sharedSecret = (String) authResponseMap.get("shared_secret");
+
+            if (auth == null || sharedSecret == null) {
+                throw new AuthorizationFailureException("Didn't receive all the fields expected " +
+                        "from the Authorizer, expected an auth and shared_secret.");
+            } else {
+                secretBoxOpener = new SecretBoxOpener(Base64.decode(sharedSecret));
+                return auth.getBytes();
+            }
+
+        } catch (final AuthorizationFailureException e) {
+            throw e; // pass this upwards
+        } catch (final Exception e) {
+            // any other errors need to be captured properly and passed upwards
+            throw new AuthorizationFailureException("Unable to parse response from Authorizer", e);
+        }
+    }
+
+    @Override
+    public String toSubscribeMessage() {
+
+        byte[] authKey = authenticate();
```

Comment:
```suggestion
        String auth = authenticate();
```
not compiling (should be applied manually)

Reply:
👍 have updated to make compile.